bundles/org.openhab.binding.openthermgateway/src/main/java/org/openhab/binding/openthermgateway/handler/BaseDeviceHandler.java

```diff
@@ -0,0 +1,164 @@
+/**
+ * Copyright (c) 2010-2021 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.openthermgateway.handler;
+
+import static org.openhab.binding.openthermgateway.internal.OpenThermGatewayBindingConstants.*;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.openhab.binding.openthermgateway.internal.DataItem;
+import org.openhab.binding.openthermgateway.internal.DataItemGroup;
+import org.openhab.binding.openthermgateway.internal.Message;
+import org.openhab.binding.openthermgateway.internal.TspFhbSizeDataItem;
+import org.openhab.binding.openthermgateway.internal.TspFhbValueDataItem;
+import org.openhab.core.thing.Bridge;
+import org.openhab.core.thing.Channel;
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.thing.Thing;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusDetail;
+import org.openhab.core.thing.binding.BaseThingHandler;
+import org.openhab.core.thing.binding.ThingHandlerCallback;
+import org.openhab.core.thing.type.ChannelKind;
+import org.openhab.core.thing.type.ChannelTypeUID;
+import org.openhab.core.types.Command;
+import org.openhab.core.types.State;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link BaseDeviceHandler} is a base class for actual Things.
+ *
+ * @author Arjen Korevaar - Initial contribution
+ */
+@NonNullByDefault
+public abstract class BaseDeviceHandler extends BaseThingHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(BaseDeviceHandler.class);
+
+    public BaseDeviceHandler(Thing thing) {
+        super(thing);
+    }
+
+    @Override
+    public void initialize() {
+        Bridge bridge = getBridge();
+
+        if (bridge == null) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.NONE, "Bridge is missing");
+        } else if (bridge.getStatus() != ThingStatus.ONLINE) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE);
```

Comment:
If you leave initialize() with at least status `UNKNOWN`, this is not needed - will be done automatically by the framework.

Reply:
Is it documented somewhere what is done automatically by the framework?

Reply:
> If you leave initialize() with at least status UNKNOWN, this is not needed

@jlaur I think your statement is not true; if you examine the code of `bridgeStatusChanged` (below) you can see that the status is only updated if the thing has ThingStatusDetail.BRIDGE_OFFLINE

```
    public void bridgeStatusChanged(ThingStatusInfo bridgeStatusInfo) {
        if (bridgeStatusInfo.getStatus() == ThingStatus.ONLINE
                && getThing().getStatusInfo().getStatusDetail() == ThingStatusDetail.BRIDGE_OFFLINE) {
            updateStatus(ThingStatus.ONLINE, ThingStatusDetail.NONE);
        } else if (bridgeStatusInfo.getStatus() == ThingStatus.OFFLINE) {
            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE);
        }
    }
```

Reply:
@fwolter - maybe you can shed a bit more light on this? Was remembering this comment: https://github.com/openhab/openhab-addons/pull/12061#discussion_r791120507. Now not sure if I introduced a bug in hdpowerview, and if more framework code is involved and will handle this. Was not able to find documentation.

Reply:
@jlaur is right. @andrewfg the status is also changed when the bridge becomes offline.

Reply:
@fwolter & @jlaur can you please explain this more? I still think your statements are wrong. The code that your comments are pointing to, is in the thing handler’s initialize() method. Yes of course the bridgeStatusChanged() in the base thing handler does change the thing state to match the bridge state, BUT IT ONLY DOES SO IF THE THING HAS ALREADY REACHED THE STATE ONLINE, OR OFFLINE/BRIDGE-OFFLINE, otherwise the thing status is NOT UPDATED, and the thing never goes online.

To be more specific, I tried removing the code that you are asking me to, AND THE BINDING DOES NOT WORK. So I will DEFINITELY NOT make the change you are asking for.

Reply:
The framework updates the ThingStatuses of the child Things only if they are ONLINE, OFFLINE or *UNKNOWN*. If you leave `initialize()` with either of these, it should work.

Reply:
> framework updates the ThingStatuses of the child Things only if they are .. UNKNOWN

@fwolter please point me to the code in the framework where it updates the state if the thing state is UNKNOWN. I have tested setting it at UNKNOWN, and the thing does NOT get updated.

> framework updates the ThingStatuses of the child Things only if they are ONLINE, OFFLINE

Exactly. It is for that exact reason that I am setting it so in the initialize() method. So why do you want me to NOT do that?

Reply:
This calls `initialize()` on the Things when the bridge is initialized: https://github.com/openhab/openhab-core/blob/87ecfade7473b9c5c05c80a521679fea2c08ad6b/bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingManagerImpl.java#L235 The actual online/offline/unknown check is here: https://github.com/openhab/openhab-core/blob/cabb3f73150d13f44d647f46af8362e396684b92/bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/util/ThingHandlerHelper.java#L37

I just tested it with the LCN binding and it works.

Reply:
@fwolter many thanks for the insights. But can you please be specific: what specifically do you want me to change in the thing’s initialize() method? It looks fine to me. And it works fine. And I don’t understand what you don’t like about it. Please help me!

Reply:
These lines can be safely removed:
```
} else if (bridge.getStatus() != ThingStatus.ONLINE) {
            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE);
```
It's just an optimization. I won't insist on changing it.

Reply:
For the posterity, this works in #12330.